ashlar/thumbnail.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import print_function, division
+import sys
+try:
+    import pathlib
+except ImportError:
+    import pathlib2 as pathlib
+import numpy as np
+from . import utils
+from skimage.transform import rescale
+from skimage.feature import register_translation
+from skimage.io import imsave
+import warnings
+
+
+def make_thumbnail(reader, channel=0, scale=0.05):
+    
+    metadata = reader.metadata
+    positions = metadata.positions - metadata.origin
+    coordinate_max = (positions + metadata.size).max(axis=0) 
+    mshape = ((coordinate_max+ 1) * scale).astype(int)
+    mosaic = np.zeros(mshape, dtype=np.uint16)
+
+    total = reader.metadata.num_images
+    for i in range(total):
+        sys.stdout.write("\r        Loading %d/%d" % (i + 1, total))
+        sys.stdout.flush()
+        utils.paste(
+            mosaic, rescale(reader.read(c=channel, series=i), scale), positions[i] * scale, np.maximum
+        )
+    print()
+
+    return mosaic
+
+
+def calculate_image_offset(img1, img2, upsample_factor=1):
+
+    ref = utils.fft2(utils.whiten(img1, 0))
+    test = utils.fft2(utils.whiten(img2, 0))
+
+    shift, error, _ = register_translation(ref, test, upsample_factor, 'fourier')
+
+    return shift
+
+
+def calculate_cycle_offset(reader1, reader2, channel=0, scale=0.05, save=(False, False)):
+    
+    img1 = reader1.thumbnail_img \
+        if hasattr (reader1, 'thumbnail_img') \
+        else make_thumbnail(reader1, channel=channel, scale=scale)
+    img2 = make_thumbnail(reader2, channel=channel, scale=scale)
+    reader1.thumbnail_img = img1
+
+    for file_path, img in zip(
+        [reader1.path, reader2.path],
+        [img1*save[0], img2*save[1]]
+    ):
+        if img.any(): _save_as_tif(img, file_path, post_fix='-thumbnail')
+    
+    if img1.shape == img2.shape:
+        img_offset = calculate_image_offset(img1, img2, int(1/scale)) * 1/scale
+    else:
+        padded_shape = np.array((img1.shape, img2.shape)).max(axis=0)
+        padded_img1, padded_img2 = np.zeros(padded_shape), np.zeros(padded_shape)
+        utils.paste(padded_img1, img1, [0, 0])
+        utils.paste(padded_img2, img2, [0, 0])
+        img_offset = calculate_image_offset(
+            padded_img1,
+            padded_img2, 
+            int(1/scale)
+        ) * 1/scale
+        
+    ori_diff = reader2.metadata.origin - reader1.metadata.origin
+
+    print(
+        '\r        Estimated offset [y x] =',
+        img_offset - ori_diff
+    )
+
+    return img_offset - ori_diff
+
+
+def thumbnail(reader, post_fix='-thumbnail', channel=0, scale=0.05):
+
+    img = make_thumbnail(reader, channel, scale)
+    _save_as_tif(img, reader.path, post_fix)
+    
+    return img
+    
+
+def _save_as_tif(img, file_path, post_fix=''):
```

Comment:
I see how it could be useful to save the thumbnail images to files, but I think implicitly writing files from deep in the utility functions like this is not an ideal design. If anything it should be managed by the top-level command line script so the user can decide if they want it. And the output location would need to be configurable as perhaps the input files live in a directory that's read-only to the user, or maybe the thumbnails would just go in the `-o` output directory. 

Maybe the solution for now is to leave the save_as_tif function here but remove all calls to it. Eventually we can re-enable it as part of the "QC" output report we've been planning.